File: captainhcg/go/270-closest-binary-search-tree-value.go
func closestValue(root *TreeNode, target float64) int {
    closest := root.Val
    for root != nil {
        if abs(float64(root.Val) - target) < abs(float64(closest) - target){
            closest = root.Val
        }
        if float64(root.Val) < target{
            root = root.Right
        } else{
            root = root.Left
        }
    }
    return closest
}

func abs(v float64) float64{
    if v < 0{
        return -v
    }
    return v
}

File: bwu/binary_tree/270-closest-binary-search-tree-value.py
class Solution(object):
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        if not root:
            return -1
            
        curr = root
        ret = float('inf')
        while curr:
            if abs(curr.val - target) < abs(ret - target):
                ret = curr.val
            if target == curr.val:
                return curr.val
            elif target < curr.val:
                curr = curr.left
            else:
                curr = curr.right
        
        return ret
